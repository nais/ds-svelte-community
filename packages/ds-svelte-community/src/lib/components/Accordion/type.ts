import type { Snippet } from "svelte";
import type { HTMLAttributes } from "svelte/elements";

export const sizes = ["large", "medium", "small"] as const;

export interface AccordionProps extends HTMLAttributes<HTMLDivElement> {
	/**
	 * The size of the accordion.
	 */
	size?: (typeof sizes)[number];

	/**
	 * Whether to indent content or not
	 */
	indent?: boolean;

	/**
	 * List of AccordionItem components.
	 */
	children: Snippet;
}

export interface AccordionItemProps extends HTMLAttributes<HTMLDivElement> {
	/**
	 * Whether the accordion item is open or not.
	 */
	open?: boolean;

	/**
	 * Heading of the accordion item.
	 */
	heading?: string | Snippet;

	/**
	 * Aria text
	 */
	showMoreText?: string;

	/**
	 * Content
	 */
	children: Snippet;
}

export type AccordionContext = {
	size: AccordionProps["size"];
};
