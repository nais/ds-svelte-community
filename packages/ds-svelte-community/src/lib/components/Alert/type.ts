import type { Snippet } from "svelte";

export const variants = ["error", "warning", "info", "success"] as const;
export const sizes = ["medium", "small"] as const;

export interface Props {
	/**
	 * Changes colors and icon usage when changed
	 */
	variant?: (typeof variants)[number];

	/**
	 * Changes padding and font-sizes
	 */
	size?: (typeof sizes)[number];

	/**
	 * Removes border-radius
	 */
	fullWidth?: boolean;

	/**
	 * Removes background from Alert
	 */
	inline?: boolean;

	/**
	 * Title attribute on the icon
	 */
	iconTitleText?: string;

	/**
	 * Alert content
	 */
	children: Snippet;

	/**
	 * Sets max-width on the content to 43.5rem.
	 */
	contentMaxWidth?: boolean;

	/**
	 * Adds a close-button (X).
	 *
	 * **Requires onClose to be set**.
	 */
	closeButton?: boolean;

	/**
	 * Icon text for screen readers used for the close button.
	 */
	closeButtonIconText?: string;

	/**
	 * Callback for alert wanting to close.
	 *
	 * **Requires closeButton to be true**.
	 */
	onClose?: () => void;

	[key: string]: unknown;
}
