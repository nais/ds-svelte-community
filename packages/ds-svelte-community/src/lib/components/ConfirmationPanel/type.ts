import type { HTMLAttributes } from "svelte/elements";

export const sizes = ["medium", "small"] as const;

export interface Props extends HTMLAttributes<HTMLDivElement> {
	error?: string;
	errorId?: string;
	size?: (typeof sizes)[number];
	value?: string;
	id?: string;
	checked?: boolean;
}
