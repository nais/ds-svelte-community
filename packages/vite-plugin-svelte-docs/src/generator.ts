import path from "path";
import type * as tsm from "ts-morph";
import { Project, ts } from "ts-morph";

import { convert } from "./svelte.js";
import type {
	Doc,
	Snippet as DocSnippet,
	Prop,
	SlotLet,
	Slots,
	SvelteEvent,
	Type,
} from "./types.js";

type ContextNode = {
	getKind(): ts.SyntaxKind;
	getText(): string;
};

class Context {
	parent?: Context;
	node: ContextNode;
	debug: boolean;
	numParents = 0;

	constructor(node: ContextNode, parent?: Context, debug: boolean = false) {
		this.node = node;
		this.parent = parent;
		this.debug = !!(debug || (parent && parent.debug));
		this.numParents = parent ? parent.numParents + 1 : 0;
	}

	public toString(): string {
		let pre = "";
		if (this.parent) {
			pre = this.parent.toString() + "\n";
		}
		pre += "  ".repeat(this.numParents);
		return pre + this.toStringWithoutParent();
	}

	public toStringWithoutParent(): string {
		switch (this.node.getKind()) {
			case ts.SyntaxKind.SourceFile:
				return "SourceFile";
		}

		return this.node.getText() + " [[" + ts.SyntaxKind[this.node.getKind()] + "]]";
	}

	public log(...args: unknown[]) {
		if (!this.debug) {
			return;
		}
		console.log(this.toString(), ...args);
	}

	public time(label: string) {
		if (!this.debug) {
			return;
		}

		console.time(label);
	}

	public timeEnd(label: string) {
		if (!this.debug) {
			return;
		}
		console.timeEnd(label);
	}
}

export class Generator {
	private project: Project;

	constructor(svelte2tsxPath: string) {
		this.project = new Project({
			compilerOptions: {
				lib: ["esnext"],
				noEmit: true,
				allowJs: true,
			},
		});

		const svelteTsxFiles = ["./svelte-shims-v4.d.ts", "./svelte-jsx-v4.d.ts"].map((f) =>
			ts.sys.resolvePath(path.resolve(path.dirname(svelte2tsxPath), f)),
		);

		this.project.addSourceFilesAtPaths(svelteTsxFiles);
	}

	renameFile(filename: string) {
		// replace .svelte with .ts to allow the file to be used in the project
		return filename.replace(/\.svelte$/, ".ts");
	}

	public addSvelteFile(filename: string, code: string) {
		filename = this.renameFile(filename);

		const dts = convert(filename, code);

		this.project.createSourceFile(filename, dts, { overwrite: true });
	}

	public docFor(filename: string, debug = false): Doc {
		filename = this.renameFile(filename);

		const sourceFile = this.project.getSourceFileOrThrow(filename);
		if (!sourceFile) {
			throw new Error(`Source file ${filename} not found`);
		}

		const ctx = new Context(sourceFile, undefined, debug);
		const ret: Doc = {
			description: "",
			props: [],
			slots: [],
			events: [],
		};

		const des = sourceFile.getDescendantsOfKind(ts.SyntaxKind.TypeAliasDeclaration);

		sourceFile.getExportSymbols().forEach((sym) => {
			// Only if default export
			if (sym.getName() !== "default") {
				return;
			}

			const description = ts.displayPartsToString(
				sym.compilerSymbol.getDocumentationComment(undefined),
			);

			ret.description = description;
		});

		ctx.time("forEach");
		des.forEach((node) => {
			const nctx = new Context(node, ctx);
			const name = node.getName();
			if (name.endsWith("Slots")) {
				nctx.time("slots");
				ret.slots = this.parseSlots(nctx, node);
				nctx.timeEnd("slots");
			} else if (name.endsWith("Props") || name.endsWith("Props_")) {
				nctx.time("props");
				ret.props = this.parseProps(nctx, node);

				ret.props = this.parsePropsDefaults(nctx, sourceFile, ret.props);

				nctx.timeEnd("props");
			} else if (name.endsWith("Events")) {
				nctx.time("events");
				ret.events = this.parseEvents(nctx, node);
				nctx.timeEnd("events");
			} else {
				nctx.log("Unknown type alias", name);
			}
		});
		ctx.timeEnd("forEach");

		if (ret.slots.length === 0) {
			ret.slots = ret.props
				.filter((p) => {
					return !Array.isArray(p.type) && p.type.type === "snippet";
				})
				.map((p) => {
					const s = p.type as DocSnippet;
					return {
						name: p.name,
						lets: s.lets,
						description: p.description,
						snippet: true,
						optional: !!p.optional,
					};
				});

			ret.events = ret.props
				.filter((p) => {
					return !Array.isArray(p.type) && p.type.type === "function";
				})
				.map((p) => {
					const ret = {
						name: p.name,
						description: p.description,
						type: p.type,
					} as SvelteEvent;
					if (p.optional !== undefined) {
						ret.optional = p.optional;
					}
					return ret;
				});

			ret.props = ret.props.filter((p) => {
				if (Array.isArray(p.type)) {
					return true;
				}

				switch (p.type.type) {
					case "snippet":
					case "function":
						return false;
				}
				return true;
			});
		} else {
			ret.props = ret.props.filter((p) => {
				if (p.name === "children" && !Array.isArray(p.type) && p.type.type === "unknown") {
					return false;
				}
				return true;
			});
		}

		const sortFn = (a: Prop | Slots | SvelteEvent, b: Prop | Slots | SvelteEvent) => {
			if (a.optional && !b.optional) {
				return 1;
			}
			if (!a.optional && b.optional) {
				return -1;
			}
			return a.name.localeCompare(b.name);
		};

		ret.props = ret.props.sort(sortFn);
		ret.slots = ret.slots.sort(sortFn);
		ret.events = ret.events.sort(sortFn);

		return ret;
	}

	parsePropsDefaults(ctx: Context, sourceFile: tsm.SourceFile, props: Prop[]): Prop[] {
		sourceFile.getChildrenOfKind(ts.SyntaxKind.FunctionDeclaration).forEach((node) => {
			node
				.getBody()
				?.getDescendantsOfKind(ts.SyntaxKind.VariableDeclarationList)
				.forEach((vd) => {
					const be = vd.getDescendantsOfKind(ts.SyntaxKind.BindingElement);
					be.forEach((v) => {
						const name = v.getNameNode().getText();
						const prop = props.find((p) => p.name === name);
						if (!prop) {
							ctx.log("Unable to find prop", name);
							return;
						}

						const initializer = v.getInitializer();
						if (!initializer) {
							ctx.log("No initializer for prop", name);
							return;
						}

						prop.default = initializer.getText();
					});

					if (be.length == 0) {
						vd.getDescendantsOfKind(ts.SyntaxKind.VariableDeclaration).forEach((v) => {
							const name = v.getName();
							const prop = props.find((p) => p.name === name);
							if (!prop) {
								ctx.log("Unable to find prop", name);
								return;
							}

							const initializer = v.getInitializer();
							if (!initializer) {
								ctx.log("No initializer for prop", name);
								return;
							}

							prop.default = initializer.getText();
						});
					}
				});
		});
		return props;
	}

	parseEvents(ctx: Context, node: tsm.TypeAliasDeclaration): SvelteEvent[] {
		const properties = node.getType().getProperties();

		return properties
			.map((prop): SvelteEvent | null => {
				const name = prop.getName();
				const description = ts.displayPartsToString(
					prop.compilerSymbol.getDocumentationComment(undefined),
				);

				if (prop.getFullyQualifiedName().includes("node_modules")) {
					return null;
				}

				return {
					name,
					description,
					optional: true,
				};
			})
			.filter((p): p is SvelteEvent => !!p);
	}

	parseSlots(ctx: Context, node: tsm.TypeAliasDeclaration) {
		const properties = node.getType().getProperties();

		return properties
			.map((prop): Slots | null => {
				const name = prop.getName();
				const description = ts.displayPartsToString(
					prop.compilerSymbol.getDocumentationComment(undefined),
				);

				if (prop.getFullyQualifiedName().includes("node_modules")) {
					return null;
				}

				// Node:
				const node = prop.getValueDeclaration();
				if (!node) {
					return null;
				}

				const nctx = new Context(node, ctx);

				const lets = node
					.getType()
					.getProperties()
					.map((prop): SlotLet | null => {
						const name = prop.getName();
						// const description = ts.displayPartsToString(
						// 	prop.compilerSymbol.getDocumentationComment(undefined),
						// );

						if (prop.getFullyQualifiedName().includes("node_modules")) {
							return null;
						}

						// Node:
						const node = prop.getValueDeclaration();
						if (!node) {
							return null;
						}
						return {
							name,
							type: this.typeOf(nctx, node),
						};
					})
					.filter((p): p is SlotLet => !!p);

				return {
					name,
					description,
					lets,
					optional: true,
				};
			})
			.filter((p): p is Slots => !!p);
	}

	parseProps(ctx: Context, node: tsm.TypeAliasDeclaration): Prop[] {
		ctx.time("getType");
		const type = node.getType();
		ctx.timeEnd("getType");
		ctx.time("getProps");
		const properties = type.getProperties();
		ctx.timeEnd("getProps");

		return properties
			.map((prop): Prop | null => {
				const name = prop.getName();
				ctx.time("prop_" + name);
				const description = ts.displayPartsToString(
					prop.compilerSymbol.getDocumentationComment(undefined),
				);

				if (prop.getFullyQualifiedName().includes("node_modules")) {
					return null;
				}

				// Node:
				const node = prop.getValueDeclaration();
				if (!node) {
					return null;
				}
				ctx.timeEnd("prop_" + name);
				return {
					name,
					type: this.typeOf(ctx, node),
					description,
					optional: prop.isOptional(),
				};
			})
			.filter((p): p is Prop => !!p);
	}

	typeOf(parentContext: Context, node: tsm.Node): Type {
		const ctx = new Context(node, parentContext);
		switch (node.getKind()) {
			case ts.SyntaxKind.PropertySignature:
				return this.typeOf(ctx, (node as tsm.PropertySignature).getTypeNodeOrThrow());
			case ts.SyntaxKind.UnionType:
				return {
					type: "union",
					values: (node as tsm.UnionTypeNode).getTypeNodes().map((t) => this.typeOf(ctx, t)),
				};
			case ts.SyntaxKind.LiteralType:
				return { type: "literal", value: (node as tsm.LiteralTypeNode).getText() };
			case ts.SyntaxKind.StringKeyword:
				return { type: "string" };
			case ts.SyntaxKind.NumberKeyword:
				return { type: "number" };
			case ts.SyntaxKind.BooleanKeyword:
				return { type: "boolean" };
			case ts.SyntaxKind.UndefinedKeyword:
				return { type: "undefined" };
			case ts.SyntaxKind.NullKeyword:
				return { type: "null" };
			case ts.SyntaxKind.FunctionType:
				return { type: "function", signature: node.getText() };
			case ts.SyntaxKind.ImportType:
				return { type: "unknown" };
			case ts.SyntaxKind.TypeReference:
				return this.typeReference(ctx, node as tsm.TypeReferenceNode);
			case ts.SyntaxKind.ImportSpecifier:
				return this.importSpecifier(ctx, node as tsm.ImportSpecifier);
			case ts.SyntaxKind.TypeAliasDeclaration:
				return this.typeOf(ctx, (node as tsm.TypeAliasDeclaration).getTypeNodeOrThrow());
			case ts.SyntaxKind.InterfaceDeclaration:
				return { type: "interface", name: (node as tsm.InterfaceDeclaration).getName() };
			case ts.SyntaxKind.PropertyAssignment:
				break;
			case ts.SyntaxKind.IndexedAccessType:
				return ((): Type => {
					const iatn = node as tsm.IndexedAccessTypeNode;
					const otn = iatn.getObjectTypeNode() as tsm.TypeReferenceNode;

					const lit = iatn.getIndexTypeNode();
					if (!lit.isKind(ts.SyntaxKind.LiteralType)) {
						if (lit.isKind(ts.SyntaxKind.NumberKeyword)) {
							return this.typeOf(ctx, otn);
						}

						ctx.log("Expected literal type");
						throw new Error("Expected literal type, got " + lit.getKindName());
					}

					const slit = lit.getLiteral();
					if (!slit.isKind(ts.SyntaxKind.StringLiteral)) {
						ctx.log("Expected string literal");
						throw new Error("Expected string literal, got " + slit.getKindName());
					}

					let nodeToUse: tsm.Node | undefined;
					otn
						.getType()
						.getProperties()
						.forEach((p) => {
							if (p.getName() === slit.getLiteralValue()) {
								nodeToUse = p.getValueDeclarationOrThrow();
							}
						});

					if (!nodeToUse) {
						ctx.log("Unable to find index");
						throw new Error("Unable to find index");
					}

					return this.typeOf(ctx, nodeToUse);
				})();
			case ts.SyntaxKind.ParenthesizedType:
				return this.parenthesizedType(ctx, node as tsm.ParenthesizedTypeNode);
			case ts.SyntaxKind.TypeQuery:
				return this.typeOf(ctx, (node as tsm.TypeQueryNode).getExprName());
			case ts.SyntaxKind.Identifier:
				return this.identifier(ctx, node as tsm.Identifier);
			case ts.SyntaxKind.VariableDeclaration:
				return this.typeOf(
					ctx,
					(node as tsm.VariableDeclaration).getInitializerOrThrow("Initializer not found"),
				);
			case ts.SyntaxKind.AsExpression:
				return this.typeOf(ctx, (node as tsm.AsExpression).getExpression());
			case ts.SyntaxKind.ArrayLiteralExpression:
				return {
					type: "union",
					values: (node as tsm.ArrayLiteralExpression)
						.getElements()
						.map((e) => this.typeOf(ctx, e)),
				};
			case ts.SyntaxKind.StringLiteral:
				return { type: "literal", value: node.getText() };
			case ts.SyntaxKind.TypeLiteral:
				ctx.log("TypeLiteral", (node as tsm.TypeLiteralNode).getMembers()[0].getText());
				return this.typeOf(ctx, (node as tsm.TypeLiteralNode).getMembers()[0]);
			// return this.typeOf((node as tsm.VariableStatement).);
		}

		console.log(" unknown kind", node.getKindName(), node.getText());
		return { type: "unknown" };
	}

	parenthesizedType(ctx: Context, node: tsm.ParenthesizedTypeNode): Type {
		const children: tsm.Node[] = [];
		node.forEachChild((c) => {
			children.push(c);
		});
		if (children.length === 1) {
			return this.typeOf(ctx, children[0]);
		}

		throw new Error("Multiple children in parenthesized type for " + node.getText());
	}

	importSpecifier(ctx: Context, node: tsm.ImportSpecifier): Type {
		let sym = node.getNameNode().getSymbolOrThrow();
		if (sym.isAlias()) {
			sym = sym.getAliasedSymbolOrThrow();
		}
		return this.typeOf(ctx, sym.getDeclarations()[0]);
	}

	identifier(ctx: Context, node: tsm.Identifier): Type {
		const sym = node.getSymbolOrThrow();
		if (sym.isAlias()) {
			return this.typeOf(ctx, sym.getAliasedSymbolOrThrow().getDeclarations()[0]);
		}
		const decl = sym.getDeclarations()[0];
		if (!decl) {
			return { type: "unknown" };
		}
		return this.typeOf(ctx, decl);
	}

	typeReference(ctx: Context, node: tsm.TypeReferenceNode): Type {
		// node.
		// const type = node.getType();
		// const symbol = type.getSymbol();
		// if (!symbol) {
		// 	return { type: "unknown" };
		// }

		// const decl = symbol.getDeclarations()[0];
		// if (!decl) {
		// 	return { type: "unknown" };
		// }

		if (node.getTypeName().getText() === "Snippet") {
			const tas = node.getTypeArguments();
			if (tas.length > 0) {
				if (tas.length > 1) {
					console.log(
						"Multiple type arguments",
						tas.map((t) => t.getText()),
					);
					throw new Error("Multiple type arguments");
				}
				const type = tas[0];
				if (type.isKind(ts.SyntaxKind.TupleType)) {
					const tuple = type as tsm.TupleTypeNode;
					const tctx = new Context(tuple, ctx);
					const types = tuple.getElements().map((e, i): SlotLet => {
						return { name: `let_${i}`, type: this.typeOf(tctx, e) };
					});
					return { type: "snippet", lets: types };
				} else {
					throw new Error("Unknown type argument");
				}
			}
			return { type: "snippet", lets: [] };
		}

		const symbol = node.getTypeName().getSymbol();
		if (!symbol) {
			console.log(" Expected symbol");
			return { type: "unknown" };
		}

		const decl = symbol.getDeclarations()[0];
		return this.typeOf(ctx, decl);
	}
}

export function typekind(typ: tsm.Type<ts.Type>) {
	if (typ.isAnonymous()) {
		return "Anonymous";
	}
	if (typ.isAny()) {
		return "Any";
	}
	if (typ.isArray()) {
		return "Array";
	}
	if (typ.isBoolean()) {
		return "Boolean";
	}
	if (typ.isBooleanLiteral()) {
		return "BooleanLiteral";
	}
	if (typ.isClass()) {
		return "Class";
	}
	if (typ.isClassOrInterface()) {
		return "ClassOrInterface";
	}
	if (typ.isEnum()) {
		return "Enum";
	}
	if (typ.isEnumLiteral()) {
		return "EnumLiteral";
	}
	if (typ.isInterface()) {
		return "Interface";
	}
	if (typ.isIntersection()) {
		return "Intersection";
	}
	if (typ.isLiteral()) {
		return "Literal";
	}
	if (typ.isNever()) {
		return "Never";
	}
	if (typ.isNull()) {
		return "Null";
	}
	if (typ.isNullable()) {
		return "Nullable";
	}
	if (typ.isNumber()) {
		return "Number";
	}
	if (typ.isNumberLiteral()) {
		return "NumberLiteral";
	}
	if (typ.isObject()) {
		return "Object";
	}
	if (typ.isReadonlyArray()) {
		return "ReadonlyArray";
	}
	if (typ.isString()) {
		return "String";
	}
	if (typ.isStringLiteral()) {
		return "StringLiteral";
	}
	if (typ.isTemplateLiteral()) {
		return "TemplateLiteral";
	}
	if (typ.isTuple()) {
		return "Tuple";
	}
	if (typ.isTypeParameter()) {
		return "TypeParameter";
	}
	if (typ.isUndefined()) {
		return "Undefined";
	}
	if (typ.isUnion()) {
		return "Union";
	}
	if (typ.isUnionOrIntersection()) {
		return "UnionOrIntersection";
	}
	if (typ.isUnknown()) {
		return "Unknown";
	}
	if (typ.isVoid()) {
		return "Void";
	}

	return "Unknown";
}
