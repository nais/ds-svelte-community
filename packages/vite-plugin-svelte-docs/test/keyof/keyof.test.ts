import { describe, expect, test } from "bun:test";
import { Generator } from "../../src/generator.js";

describe("keyof", () => {
	test("inline", async () => {
		const filename = "asdf.svelte";
		const code = `<script lang="ts">
    interface Elements {
      a: string;
      div: string;
    }

    type Props = {
      as: keyof Elements;
    };

    let props : Props = $props();
  </script>
  `;

		const gen = new Generator(require.resolve("svelte2tsx"));
		await gen.setup();
		gen.addSvelteFile(filename, code);

		const { doc } = gen.docFor(filename, false);
		expect(doc).toMatchSnapshot();
	});

	test("svelte elements", async () => {
		const filename = "asdf.svelte";
		const code = `<script lang="ts">
    import type { SvelteHTMLElements } from "svelte/elements";

    type Props = {
      as: keyof SvelteHTMLElements;
    };

    let props : Props = $props();
  </script>
  `;

		const gen = new Generator(require.resolve("svelte2tsx"));
		await gen.setup();
		gen.addSvelteFile(filename, code);

		const { doc } = gen.docFor(filename, false);
		expect(doc).toMatchSnapshot();
	});
});
