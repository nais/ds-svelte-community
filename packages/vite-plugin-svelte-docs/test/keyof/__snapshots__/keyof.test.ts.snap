// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`keyof inline 1`] = `
{
  "description": "",
  "events": [],
  "name": "asdf",
  "props": [
    {
      "description": "",
      "name": "as",
      "optional": false,
      "type": {
        "type": "union",
        "values": [
          {
            "type": "literal",
            "value": ""a"",
          },
          {
            "type": "literal",
            "value": ""div"",
          },
        ],
      },
    },
  ],
  "slots": [],
}
`;

exports[`keyof svelte elements 1`] = `
{
  "description": "",
  "events": [],
  "name": "asdf",
  "props": [
    {
      "description": "",
      "name": "as",
      "optional": false,
      "type": {
        "type": "union",
        "values": [
          {
            "type": "literal",
            "value": ""a"",
          },
          {
            "type": "literal",
            "value": ""abbr"",
          },
          {
            "type": "literal",
            "value": ""address"",
          },
          {
            "type": "literal",
            "value": ""area"",
          },
          {
            "type": "literal",
            "value": ""article"",
          },
          {
            "type": "literal",
            "value": ""aside"",
          },
          {
            "type": "literal",
            "value": ""audio"",
          },
          {
            "type": "literal",
            "value": ""b"",
          },
          {
            "type": "literal",
            "value": ""base"",
          },
          {
            "type": "literal",
            "value": ""bdi"",
          },
          {
            "type": "literal",
            "value": ""bdo"",
          },
          {
            "type": "literal",
            "value": ""big"",
          },
          {
            "type": "literal",
            "value": ""blockquote"",
          },
          {
            "type": "literal",
            "value": ""body"",
          },
          {
            "type": "literal",
            "value": ""br"",
          },
          {
            "type": "literal",
            "value": ""button"",
          },
          {
            "type": "literal",
            "value": ""canvas"",
          },
          {
            "type": "literal",
            "value": ""caption"",
          },
          {
            "type": "literal",
            "value": ""cite"",
          },
          {
            "type": "literal",
            "value": ""code"",
          },
          {
            "type": "literal",
            "value": ""col"",
          },
          {
            "type": "literal",
            "value": ""colgroup"",
          },
          {
            "type": "literal",
            "value": ""data"",
          },
          {
            "type": "literal",
            "value": ""datalist"",
          },
          {
            "type": "literal",
            "value": ""dd"",
          },
          {
            "type": "literal",
            "value": ""del"",
          },
          {
            "type": "literal",
            "value": ""details"",
          },
          {
            "type": "literal",
            "value": ""dfn"",
          },
          {
            "type": "literal",
            "value": ""dialog"",
          },
          {
            "type": "literal",
            "value": ""div"",
          },
          {
            "type": "literal",
            "value": ""dl"",
          },
          {
            "type": "literal",
            "value": ""dt"",
          },
          {
            "type": "literal",
            "value": ""em"",
          },
          {
            "type": "literal",
            "value": ""embed"",
          },
          {
            "type": "literal",
            "value": ""fieldset"",
          },
          {
            "type": "literal",
            "value": ""figcaption"",
          },
          {
            "type": "literal",
            "value": ""figure"",
          },
          {
            "type": "literal",
            "value": ""footer"",
          },
          {
            "type": "literal",
            "value": ""form"",
          },
          {
            "type": "literal",
            "value": ""h1"",
          },
          {
            "type": "literal",
            "value": ""h2"",
          },
          {
            "type": "literal",
            "value": ""h3"",
          },
          {
            "type": "literal",
            "value": ""h4"",
          },
          {
            "type": "literal",
            "value": ""h5"",
          },
          {
            "type": "literal",
            "value": ""h6"",
          },
          {
            "type": "literal",
            "value": ""head"",
          },
          {
            "type": "literal",
            "value": ""header"",
          },
          {
            "type": "literal",
            "value": ""hgroup"",
          },
          {
            "type": "literal",
            "value": ""hr"",
          },
          {
            "type": "literal",
            "value": ""html"",
          },
          {
            "type": "literal",
            "value": ""i"",
          },
          {
            "type": "literal",
            "value": ""iframe"",
          },
          {
            "type": "literal",
            "value": ""img"",
          },
          {
            "type": "literal",
            "value": ""input"",
          },
          {
            "type": "literal",
            "value": ""ins"",
          },
          {
            "type": "literal",
            "value": ""kbd"",
          },
          {
            "type": "literal",
            "value": ""keygen"",
          },
          {
            "type": "literal",
            "value": ""label"",
          },
          {
            "type": "literal",
            "value": ""legend"",
          },
          {
            "type": "literal",
            "value": ""li"",
          },
          {
            "type": "literal",
            "value": ""link"",
          },
          {
            "type": "literal",
            "value": ""main"",
          },
          {
            "type": "literal",
            "value": ""map"",
          },
          {
            "type": "literal",
            "value": ""mark"",
          },
          {
            "type": "literal",
            "value": ""menu"",
          },
          {
            "type": "literal",
            "value": ""menuitem"",
          },
          {
            "type": "literal",
            "value": ""meta"",
          },
          {
            "type": "literal",
            "value": ""meter"",
          },
          {
            "type": "literal",
            "value": ""nav"",
          },
          {
            "type": "literal",
            "value": ""noscript"",
          },
          {
            "type": "literal",
            "value": ""object"",
          },
          {
            "type": "literal",
            "value": ""ol"",
          },
          {
            "type": "literal",
            "value": ""optgroup"",
          },
          {
            "type": "literal",
            "value": ""option"",
          },
          {
            "type": "literal",
            "value": ""output"",
          },
          {
            "type": "literal",
            "value": ""p"",
          },
          {
            "type": "literal",
            "value": ""param"",
          },
          {
            "type": "literal",
            "value": ""picture"",
          },
          {
            "type": "literal",
            "value": ""pre"",
          },
          {
            "type": "literal",
            "value": ""progress"",
          },
          {
            "type": "literal",
            "value": ""q"",
          },
          {
            "type": "literal",
            "value": ""rp"",
          },
          {
            "type": "literal",
            "value": ""rt"",
          },
          {
            "type": "literal",
            "value": ""ruby"",
          },
          {
            "type": "literal",
            "value": ""s"",
          },
          {
            "type": "literal",
            "value": ""samp"",
          },
          {
            "type": "literal",
            "value": ""slot"",
          },
          {
            "type": "literal",
            "value": ""script"",
          },
          {
            "type": "literal",
            "value": ""section"",
          },
          {
            "type": "literal",
            "value": ""select"",
          },
          {
            "type": "literal",
            "value": ""small"",
          },
          {
            "type": "literal",
            "value": ""source"",
          },
          {
            "type": "literal",
            "value": ""span"",
          },
          {
            "type": "literal",
            "value": ""strong"",
          },
          {
            "type": "literal",
            "value": ""style"",
          },
          {
            "type": "literal",
            "value": ""sub"",
          },
          {
            "type": "literal",
            "value": ""summary"",
          },
          {
            "type": "literal",
            "value": ""sup"",
          },
          {
            "type": "literal",
            "value": ""table"",
          },
          {
            "type": "literal",
            "value": ""template"",
          },
          {
            "type": "literal",
            "value": ""tbody"",
          },
          {
            "type": "literal",
            "value": ""td"",
          },
          {
            "type": "literal",
            "value": ""textarea"",
          },
          {
            "type": "literal",
            "value": ""tfoot"",
          },
          {
            "type": "literal",
            "value": ""th"",
          },
          {
            "type": "literal",
            "value": ""thead"",
          },
          {
            "type": "literal",
            "value": ""time"",
          },
          {
            "type": "literal",
            "value": ""title"",
          },
          {
            "type": "literal",
            "value": ""tr"",
          },
          {
            "type": "literal",
            "value": ""track"",
          },
          {
            "type": "literal",
            "value": ""u"",
          },
          {
            "type": "literal",
            "value": ""ul"",
          },
          {
            "type": "literal",
            "value": ""var"",
          },
          {
            "type": "literal",
            "value": ""video"",
          },
          {
            "type": "literal",
            "value": ""wbr"",
          },
          {
            "type": "literal",
            "value": ""webview"",
          },
          {
            "type": "literal",
            "value": ""svg"",
          },
          {
            "type": "literal",
            "value": ""animate"",
          },
          {
            "type": "literal",
            "value": ""animateMotion"",
          },
          {
            "type": "literal",
            "value": ""animateTransform"",
          },
          {
            "type": "literal",
            "value": ""circle"",
          },
          {
            "type": "literal",
            "value": ""clipPath"",
          },
          {
            "type": "literal",
            "value": ""defs"",
          },
          {
            "type": "literal",
            "value": ""desc"",
          },
          {
            "type": "literal",
            "value": ""ellipse"",
          },
          {
            "type": "literal",
            "value": ""feBlend"",
          },
          {
            "type": "literal",
            "value": ""feColorMatrix"",
          },
          {
            "type": "literal",
            "value": ""feComponentTransfer"",
          },
          {
            "type": "literal",
            "value": ""feComposite"",
          },
          {
            "type": "literal",
            "value": ""feConvolveMatrix"",
          },
          {
            "type": "literal",
            "value": ""feDiffuseLighting"",
          },
          {
            "type": "literal",
            "value": ""feDisplacementMap"",
          },
          {
            "type": "literal",
            "value": ""feDistantLight"",
          },
          {
            "type": "literal",
            "value": ""feDropShadow"",
          },
          {
            "type": "literal",
            "value": ""feFlood"",
          },
          {
            "type": "literal",
            "value": ""feFuncA"",
          },
          {
            "type": "literal",
            "value": ""feFuncB"",
          },
          {
            "type": "literal",
            "value": ""feFuncG"",
          },
          {
            "type": "literal",
            "value": ""feFuncR"",
          },
          {
            "type": "literal",
            "value": ""feGaussianBlur"",
          },
          {
            "type": "literal",
            "value": ""feImage"",
          },
          {
            "type": "literal",
            "value": ""feMerge"",
          },
          {
            "type": "literal",
            "value": ""feMergeNode"",
          },
          {
            "type": "literal",
            "value": ""feMorphology"",
          },
          {
            "type": "literal",
            "value": ""feOffset"",
          },
          {
            "type": "literal",
            "value": ""fePointLight"",
          },
          {
            "type": "literal",
            "value": ""feSpecularLighting"",
          },
          {
            "type": "literal",
            "value": ""feSpotLight"",
          },
          {
            "type": "literal",
            "value": ""feTile"",
          },
          {
            "type": "literal",
            "value": ""feTurbulence"",
          },
          {
            "type": "literal",
            "value": ""filter"",
          },
          {
            "type": "literal",
            "value": ""foreignObject"",
          },
          {
            "type": "literal",
            "value": ""g"",
          },
          {
            "type": "literal",
            "value": ""image"",
          },
          {
            "type": "literal",
            "value": ""line"",
          },
          {
            "type": "literal",
            "value": ""linearGradient"",
          },
          {
            "type": "literal",
            "value": ""marker"",
          },
          {
            "type": "literal",
            "value": ""mask"",
          },
          {
            "type": "literal",
            "value": ""metadata"",
          },
          {
            "type": "literal",
            "value": ""mpath"",
          },
          {
            "type": "literal",
            "value": ""path"",
          },
          {
            "type": "literal",
            "value": ""pattern"",
          },
          {
            "type": "literal",
            "value": ""polygon"",
          },
          {
            "type": "literal",
            "value": ""polyline"",
          },
          {
            "type": "literal",
            "value": ""radialGradient"",
          },
          {
            "type": "literal",
            "value": ""rect"",
          },
          {
            "type": "literal",
            "value": ""stop"",
          },
          {
            "type": "literal",
            "value": ""switch"",
          },
          {
            "type": "literal",
            "value": ""symbol"",
          },
          {
            "type": "literal",
            "value": ""text"",
          },
          {
            "type": "literal",
            "value": ""textPath"",
          },
          {
            "type": "literal",
            "value": ""tspan"",
          },
          {
            "type": "literal",
            "value": ""use"",
          },
          {
            "type": "literal",
            "value": ""view"",
          },
          {
            "type": "literal",
            "value": ""'svelte:window'"",
          },
          {
            "type": "literal",
            "value": ""'svelte:document'"",
          },
          {
            "type": "literal",
            "value": ""'svelte:body'"",
          },
          {
            "type": "literal",
            "value": ""'svelte:fragment'"",
          },
          {
            "type": "literal",
            "value": ""'svelte:options'"",
          },
          {
            "type": "literal",
            "value": ""'svelte:head'"",
          },
        ],
      },
    },
  ],
  "slots": [],
}
`;
